src/jgikbase/idmapping/storage/mongo/id_mapping_mongo_storage.py

```diff
@@ -44,29 +45,79 @@ def __init__(self, db: Database) -> None:
         self._ensure_indexes()
 
     def _ensure_indexes(self):
-        self.db[_COL_USERS].create_index([(_FLD_AUTHSOURCE, 1), (_FLD_USER, 1)], unique=True)
+        self.db[_COL_USERS].create_index(_FLD_USER, unique=True)
         self.db[_COL_USERS].create_index(_FLD_TOKEN, unique=True)
 
-    def create_or_update_local_user(self, user: User, token: HashedToken) -> None:
+    def create_local_user(self, user: User, token: HashedToken) -> None:
+        self._check_user_inputs(user, token)
+        try:
+            self.db[_COL_USERS].insert_one({_FLD_USER: user.username,
+                                            _FLD_TOKEN: token.token_hash})
+        except DuplicateKeyError as e:
+            coll, index = self._get_duplicate_location(e)
+            if coll == _COL_USERS:
+                if index == _FLD_USER + '_1':
+                    # TODO EXCEP make duplicate user exception
+                    raise ValueError('Duplicate user: ' + user.username)
+                elif index == _FLD_TOKEN + '_1':
+                    # TODO EXCEP make duplicate token exception
+                    raise ValueError('The provided token already exists in the database')
+            # this is impossible to test
+            raise ValueError('Unexpected duplicate key exception')
+        except PyMongoError as e:
+            # TODO EXCEP handle other mongo errors
+            raise e
+
+    # this regex is gross, but matches duplicate key error text across mongo versions 2 & 3 at
+    # least.
+    _DUPLICATE_KEY_REGEX = re.compile('duplicate key error (index|collection): ' +
```

Comment:
Are you going to need this duplicate detection thing for other database resources, such as namespaces and objects? Maybe it should go in its own module?

Reply:
I think it always helps with regexes to make a comment with example strings that the regex successfully matches.

Reply:
> Maybe it should go in its own module?

I mean, it could, but given it's tightly coupled to mongodb this seems like the right place for it - there's no reason to expose it anywhere else AFAICT.

> I think it always helps with regexes to make a comment with example strings that the regex successfully matches.

I've added a todo to add this in a future PR.

Reply:
So is the idea with the regex here that mongo gives a plain-string error message and you are pulling out just the field names? If that is right, then it might be worth adding another documentation note about what's going on there.

Reply:
right, and will do